README
Mojolicious-Plugin-Cache-Lite

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

COPYRIGHT AND LICENSE

Copyright (C) 2011 sugama

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.



Changes
Revision history for Mojolicious-Plugin-CacheLite

0.03 2011/07/13
  - Performance for cache setter improved.

0.02 2011/07/13
  - Cache size measuring method changed for performance.

0.01 2011/07/11
  - initial release


t/lib/book/lib/Book.pm
package Book;

use strict;
use base 'Mojolicious';
use File::Spec::Functions;

# This method will run once at server start
sub startup {
    my $self = shift;
    my $r = $self->routes;
    $self->plugin('cache-lite');
    my $books = $r->route('/books')->to('controller-cache#books');
}

1;


t/lib/user/lib/User.pm
package User;

use strict;
use base 'Mojolicious';
use File::Spec::Functions;

# This method will run once at server start
sub startup {
    my $self = shift;

    # Routes
    my $r = $self->routes;

    $self->plugin('cache-lite');

    $r->route('/user')->via('post')->to('controller-cache#add');
    $r->route('/user/:id')->via('delete')->to('controller-cache#remove_user');
    my $books
        = $r->waypoint('/user')->via('get')->to('controller-cache#users');
    my $more
        = $books->waypoint('/:id')->via('get')->to('controller-cache#show');
    $more->route('/email')->via('get')->to('controller-cache#email');
    $more->route('/name')->via('get')->to('controller-cache#name');
}

1;


t/lib/MojoliciousTest.pm
package MojoliciousTest;
use Mojo::Base 'Mojolicious';

sub development_mode {
  my $self = shift;

  # Static root for development
  $self->static->root($self->home->rel_dir('public_dev'));
}

# "Let's face it, comedy's a dead art form. Tragedy, now that's funny."
sub startup {
  my $self = shift;
  
  $self->plugin(cache_lite => {keygen => sub{
    my $c = shift;
    my $path = $c->req->url->path;
    if ($path =~ qr{/auth/}) {
      return;
    }
    if ($c->req->headers->header('X-Pass')) {
      return;
    }
    if ($c->req->headers->header('If-Modified-Since')) {
      return;
    }
    return $path;
  }});

  # Plugin
  unshift @{$self->plugins->namespaces},
    $self->routes->namespace . '::Plugin';
  $self->plugin('test_plugin');

  # Templateless renderer
  $self->renderer->add_handler(
    test => sub {
      my ($self, $c, $output) = @_;
      $$output = 'Hello Mojo from a templateless renderer!';
    }
  );

  # Renderer for a different file extension
  $self->renderer->add_handler(xpl => $self->renderer->handlers->{epl});

  # Shortcut for "/fun*" routes
  $self->routes->add_shortcut(
    fun => sub {
      my ($r, $append) = @_;
      $r->route("/fun$append");
    }
  );

  # Session domain
  $self->sessions->cookie_domain('.example.com');

  # Routes
  my $r = $self->routes;

  # /exceptional/*
  $r->route('/exceptional/:action')->to('exceptional#');

  # /exceptional_too/*
  $r->bridge('/exceptional_too')->to('exceptional#this_one_might_die')
    ->route('/:action');

  # /fun/time
  $r->fun('/time')->to('foo#fun');

  # /happy/fun/time
  $r->route('/happy')->fun('/time')->to('foo#fun');

  # /auth (authentication bridge)
  my $auth = $r->bridge('/auth')->to(
    cb => sub {
      return 1 if shift->req->headers->header('X-Bender');
      return;
    }
  );

  # /auth/authenticated
  $auth->route('/authenticated')->to('foo#authenticated');

  # /stash_config
  $r->route('/stash_config')
    ->to(controller => 'foo', action => 'config', config => {test => 123});

  # /test4 - named route for url_for
  $r->route('/test4/:something')->to('foo#something', something => 23)
    ->name('something');

  # /somethingtest - refer to another route with url_for
  $r->route('/somethingtest')->to('foo#something');

  # /something_missing - refer to a non existing route with url_for
  $r->route('/something_missing')->to('foo#url_for_missing');

  # /test3 - no class, just a namespace
  $r->route('/test3')
    ->to(namespace => 'MojoliciousTestController', method => 'index');

  # /test2 - different namespace test
  $r->route('/test2')->to(
    namespace => 'MojoliciousTest2',
    class     => 'Foo',
    method    => 'test'
  );

  # /test5 - only namespace test
  $r->route('/test5')->to(
    namespace => 'MojoliciousTest2::Foo',
    method    => 'test'
  );

  # /test6 - no namespace test
  $r->route('/test6')->to(
    namespace  => '',
    controller => 'mojolicious_test2-foo',
    action     => 'test'
  );

  # /withblock - template with blocks
  $r->route('/withblock')->to('foo#withblock');

  # /staged - authentication with bridges
  my $b = $r->bridge('/staged')->to(controller => 'foo', action => 'stage1');
  $b->route->to(action => 'stage2');

  # /shortcut/act
  # /shortcut/ctrl
  # /shortcut/ctrl-act - shortcuts to controller#action
  $r->route('/shortcut/ctrl-act')
    ->to('foo#config', config => {test => 'ctrl-act'});
  $r->route('/shortcut/ctrl')
    ->to('foo#', action => 'config', config => {test => 'ctrl'});
  $r->route('/shortcut/act')
    ->to('#config', controller => 'foo', config => {test => 'act'});

  # /foo/session - session cookie with domain
  $r->route('/foo/session')->to('foo#session_domain');

  # /rss.xml - mixed formats
  $r->route('/rss.xml')->to('foo#bar', format => 'rss');

  # /*/* - the default route
  $r->route('/(controller)/(action)')->to(action => 'index');
}

1;


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

my %params = (
    NAME          => 'Mojolicious::Plugin::CacheLite',
    AUTHOR        => 'sugama <sugama@jamadam.com>',
    VERSION_FROM  => 'lib/Mojolicious/Plugin/CacheLite.pm',
    ABSTRACT_FROM => 'lib/Mojolicious/Plugin/CacheLite.pm',
    LICENSE       => 'perl',
    PREREQ_PM     => {
    },
    BUILD_REQUIRES => {
        'Test::More'          => '0.47',
        'Test::UseAllModules' => '0.10',
        'Mojolicious'    	  => '1.59',
    },
    META_MERGE => {
        resources => {
            repository => '',
        },
    },
);

my $eumm = $ExtUtils::MakeMaker::VERSION;
delete $params{LICENSE}          if $eumm < 6.31;
delete $params{MIN_PERL_VERSION} if $eumm < 6.48;
delete $params{META_MERGE}       if $eumm < 6.46;
delete $params{META_ADD}         if $eumm < 6.46;
delete $params{LICENSE}          if $eumm < 6.31;

if ($eumm < 6.52 && $params{CONFIGURE_REQUIRES}) {
    $params{PREREQ_PM} = {
        %{ $params{PREREQ_PM}          || {} },
        %{ $params{CONFIGURE_REQUIRES} },
    };
    delete $params{CONFIGURE_REQUIRES};
}
if ($eumm < 6.5503 && $params{BUILD_REQUIRES}) {
    $params{PREREQ_PM} = {
        %{ $params{PREREQ_PM}      || {} },
        %{ $params{BUILD_REQUIRES} },
    };
    delete $params{BUILD_REQUIRES};
}

WriteMakefile(%params);


lib/Mojo/Cache/Extended.pm
package Mojo::Cache::Extended;
use strict;
use warnings;
use Mojo::Base -base;
    
    has 'max_keys' => 100;
    has 'max_size' => 5000000;
    has 'size_of';
    
    my $ATTR_CACHE      = 1;
    my $ATTR_STACK      = 2;
    my $ATTR_TOTAL      = 3;
    
    sub get {
        if (my $cache = $_[0]->{$ATTR_CACHE}->{$_[1]}) {
            if ($cache->[3]) {
                for my $code (@{$cache->[3]}) {
                    if ($code->($cache->[1])) {
                        return;
                    }
                }
            }
            $cache->[0];
        }
    }
    
    sub set {
        my ($self, $key, $value) = @_;
        
        my $keys  = $self->max_keys;
        my $max_size = $self->max_size;
        my $cache = $self->{$ATTR_CACHE} ||= {};
        my $stack = $self->{$ATTR_STACK} ||= [];
        $self->{$ATTR_TOTAL} ||= 0;
        my $length;
        if (my $size_of = $self->size_of) {
            $length = $size_of->($value);
            $self->{$ATTR_TOTAL} += $length;
        }
        
        while (@$stack >= $keys || $self->{$ATTR_TOTAL} > $max_size) {
            my $key = shift @$stack;
            $self->{$ATTR_TOTAL} -= $cache->{$key}->[2] || 0;
            delete $cache->{$key};
        }
        
        push @$stack, $key;
        $cache->{$key} = [$value, time, $length, undef];
        
        return $self;
    }
    
    sub set_expire {
        my ($self, $key, $cb) = @_;
        push(@{$self->{$ATTR_CACHE}->{$key}->[3]}, $cb);
    }

1;
__END__

=head1 NAME

Mojo::Cache::Extended - 

=head1 SYNOPSIS

=head1 DESCRIPTION

L<Mojo::Cache::Extended> class represents caches. 

Note that this module is EXPERIMENTAL and might change without warning!

=head1 ATTRIBUTES

L<Mojo::Cache> implements the following attributes.

=head2 C<max_keys>

  my $max_keys = $cache->max_keys;
  $cache       = $cache->max_keys(50);

Maximum number of cache keys, defaults to C<100>.

=head2 C<max_size>

  my $max_size = $cache->max_size;
  $cache       = $cache->max_size(10000000);

Maximum size of value length sum, defaults to C<5000000>.

=head2 C<size_of>

  $cache->size_of(sub {length(shift)});

This attribute specifies the rule for measure the size of cache.

=head1 METHODS

L<Mojo::Cache> inherits all methods from L<Mojo::Base> and implements the
following new ones.

=head2 C<get>

  my $value = $cache->get('foo');

Get cached value.

=head2 C<set>

  $cache = $cache->set(foo => 'bar');

Set cached value.

=head2 C<set_expire>

Sets expiration condition by code refs. This method always wrap
the original one.
    
    $cache->set_expire('foo' => sub{
        my $ts = shift;
    });

=head2 C<guess_size_of>

This is aimed at internal use. This wild guesses the object size.

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Guides>, L<http://mojolicio.us>.

=cut


lib/Mojo/Cache/Response.pm
package Mojo::Cache::Response;
use strict;
use warnings;
use Mojo::Base -base;
    
    has 'max_keys' => 100;
    has 'max_bytes' => 5000000;
    
    my $ATTR_CACHE = 1;
    my $ATTR_STACK = 2;
    my $ATTR_TOTAL = 3;
    my $ATTR_EXPIRE = 4;
    my $ATTR_TIMESTAMP = 5;
    
    sub get {
        if ($_[0]->{$ATTR_EXPIRE}) {
            my $expire = $_[0]->{$ATTR_EXPIRE}->{$_[1]};
            if ($expire && $expire->($_[0]->{$ATTR_TIMESTAMP}->{$_[1]})) {
                return;
            }
        }
        ($_[0]->{$ATTR_CACHE} || {})->{$_[1]};
    }
    
    sub set {
        my ($self, $key, $value) = @_;
        
        my $keys  = $self->max_keys;
        my $max_length = $self->max_bytes;
        my $cache = $self->{$ATTR_CACHE} ||= {};
        my $stack = $self->{$ATTR_STACK} ||= [];
        my $ts = $self->{$ATTR_TIMESTAMP} ||= {};
        $self->{$ATTR_TOTAL} ||= 0;
        $self->{$ATTR_TOTAL} += length($value);
        
        while (@$stack >= $keys || $self->{$ATTR_TOTAL} > $max_length) {
            $self->remove(shift @$stack);
        }
        
        push @$stack, $key;
        $cache->{$key} = $value;
        $ts->{$key} = time;
        
        return $self;
    }
    
    sub remove {
        my ($self, $key) = @_;
        my $cache = $self->{$ATTR_CACHE} || {};
        my $stack = $self->{$ATTR_STACK} || [];
        my $ts = $self->{$ATTR_TIMESTAMP} ||= {};
        if (defined $cache->{$key}) {
            $self->{$ATTR_TOTAL} -= length($cache->{$key});
            @$stack = grep {$_ ne $key} @$stack;
            delete $cache->{$key};
            delete $ts->{$key};
        }
    }
    
    sub set_expire {
        my ($self, $key, $cb) = @_;
        my $org = $self->{$ATTR_EXPIRE}->{$key} || sub{};
        $self->{$ATTR_EXPIRE}->{$key} = sub {
            if ($cb->(@_) || $org->(@_)) {
                $self->remove($key);
                return 1;
            }
        };
    }

1;
__END__

=head1 NAME

Mojo::Cache::Response - 

=head1 SYNOPSIS

=head1 DESCRIPTION

L<Mojo::Cache::Response> class represents caches. 

Note that this module is EXPERIMENTAL and might change without warning!

=head1 ATTRIBUTES

L<Mojo::Cache> implements the following attributes.

=head2 C<max_keys>

  my $max_keys = $cache->max_keys;
  $cache       = $cache->max_keys(50);

Maximum number of cache keys, defaults to C<100>.

=head2 C<max_bytes>

  my $max_bytes = $cache->max_bytes;
  $cache       = $cache->max_bytes(10000000);

Maximum size of value length sum, defaults to C<5000000>.

=head1 METHODS

L<Mojo::Cache> inherits all methods from L<Mojo::Base> and implements the
following new ones.

=head2 C<get>

  my $value = $cache->get('foo');

Get cached value.

=head2 C<set>

  $cache = $cache->set(foo => 'bar');

Set cached value.

=head2 C<remove>

  $cache = $cache->remove('foo');

Remove cached value.

=head2 C<set_expire>

Sets expiration condition by code refs. This method always wrap
the original one.
    
    $cache->set_expire('foo' => sub{
        my $ts = shift;
    });

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Guides>, L<http://mojolicio.us>.

=cut


lib/Mojo/Cache/ByteLimited.pm
package Mojo::Cache::ByteLimited;
use strict;
use warnings;
use Mojo::Base -base;
    
    has 'max_keys' => 100;
    has 'max_bytes' => 5000000;
    
    my $ATTR_CACHE = 1;
    my $ATTR_STACK = 2;
    my $ATTR_TOTAL = 3;
    my $ATTR_EXPIRE = 4;
    my $ATTR_TIMESTAMP = 5;
    
    sub get {
        my ($self, $key) = @_;
        if (my $val = ($self->{$ATTR_CACHE} || {})->{$key}) {
            my $expire = ($self->{$ATTR_EXPIRE} || {})->{$key};
            if ($expire && $expire->($self->{$ATTR_TIMESTAMP}->{$key})) {
                return;
            }
            return $val;
        }
    }
    
    sub set {
        my ($self, $key, $value) = @_;
        
        my $keys  = $self->max_keys;
        my $max_length = $self->max_bytes;
        my $cache = $self->{$ATTR_CACHE} ||= {};
        my $stack = $self->{$ATTR_STACK} ||= [];
        my $ts = $self->{$ATTR_TIMESTAMP} ||= {};
        $self->{$ATTR_TOTAL} ||= 0;
        $self->{$ATTR_TOTAL} += length($value);
        
        while (@$stack >= $keys || $self->{$ATTR_TOTAL} > $max_length) {
            $self->remove(shift @$stack);
        }
        
        push @$stack, $key;
        $cache->{$key} = $value;
        $ts->{$key} = time;
        
        return $self;
    }
    
    sub remove {
        my ($self, $key) = @_;
        my $cache = $self->{$ATTR_CACHE} || {};
        my $stack = $self->{$ATTR_STACK} || [];
        my $ts = $self->{$ATTR_TIMESTAMP} ||= {};
        if (defined $cache->{$key}) {
            $self->{$ATTR_TOTAL} -= length($cache->{$key});
            @$stack = grep {$_ ne $key} @$stack;
            delete $cache->{$key};
            delete $ts->{$key};
        }
    }
    
    sub set_expire {
        my ($self, $key, $cb) = @_;
        my $org = $self->{$ATTR_EXPIRE}->{$key} || sub{};
        $self->{$ATTR_EXPIRE}->{$key} = sub {
            if ($cb->(@_) || $org->(@_)) {
                $self->remove($key);
                return 1;
            }
        };
    }

1;
__END__

=head1 NAME

Mojo::Cache::ByteLimited - 

=head1 SYNOPSIS

=head1 DESCRIPTION

L<Mojo::Cache::ByteLimited> is a Mojo::Cache sub class with byte limiter.

Note that this module is EXPERIMENTAL and might change without warning!

=head1 ATTRIBUTES

L<Mojo::Cache> implements the following attributes.

=head2 C<max_keys>

  my $max_keys = $cache->max_keys;
  $cache       = $cache->max_keys(50);

Maximum number of cache keys, defaults to C<100>.

=head2 C<max_bytes>

  my $max_bytes = $cache->max_bytes;
  $cache       = $cache->max_bytes(10000000);

Maximum size of value length sum, defaults to C<5000000>.

=head1 METHODS

L<Mojo::Cache> inherits all methods from L<Mojo::Base> and implements the
following new ones.

=head2 C<get>

  my $value = $cache->get('foo');

Get cached value.

=head2 C<set>

  $cache = $cache->set(foo => 'bar');

Set cached value.

=head2 C<remove>

  $cache = $cache->remove('foo');

Remove cached value.

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Guides>, L<http://mojolicio.us>.

=cut


lib/Mojolicious/Plugin/Cache/Lite.pm
package Mojolicious::Plugin::Cache::Lite;
use strict;
use warnings;
our $VERSION = '0.01';
use base qw/Mojolicious::Plugin/;
use Mojo::JSON;
use Mojo::Cache::Extended;

    our $_EXPIRE_CODE_ARRAY = [];
    
    sub set_expire {
        my ($class, $code) = @_;
        push @$_EXPIRE_CODE_ARRAY, $code;
    }
    
    sub register {
        my ( $self, $app, $conf ) = @_;
        
        my $key_generater = $conf->{key_generater} || sub {
            shift->req->url->to_abs->to_string;
        };
        
        my $cache = Mojo::Cache::Extended->new();
        
        if ($conf->{max_bytes}) {
            $cache->max_bytes($conf->{max_bytes});
        }
        
        my $on_process_org = $app->on_process;
        
        $app->on_process(sub {
            
            my ($app, $c) = @_;
            
            my $active = ($c->req->method eq 'GET');
            my $key = $key_generater->($c);
            
            if ($active && $key) {
                my $data = Mojo::JSON->decode($cache->get($key));
                if (defined $data) {
                    $app->log->debug("serving from cache for $key");
                    $c->res->code($data->{code});
                    $c->res->headers(bless $data->{headers}, 'Mojo::Headers');
                    $c->res->body($data->{body});
                    $c->rendered;
                    return;
                }
            }
            
            local $_EXPIRE_CODE_ARRAY;
            
            $on_process_org->($app, $c);
            
            my $code = $c->res->code;
            
            if ($active && $key && $code && $code == 200) {
                $app->log->debug("storing in cache for $key");
                my %header = %{$c->res->headers};
                $cache->set($key, Mojo::JSON->encode({
                    body    => $c->res->body,
                    headers => \%header,
                    code    => $c->res->code,
                }));
                for my $code (@$_EXPIRE_CODE_ARRAY) {
                    $cache->set_expire($key, $code);
                }
            }
        });
    }

1;

__END__

=head1 NAME

Mojolicious::Plugin::Cache::Lite - On memory cache plugin [ALPHA]

=head1 SYNOPSIS

    use Mojolicious::Plugin::Cache::Lite;
    
    sub startup {
        my $self = shift;
        
        $self->plugin('cache_lite');
        
        or
        
        $self->plugin(cache_lite => {
            max_bytes => 1000000,
            key_generater => sub {
                my $c = shift;
                
                # generate key here maybe with $c
                # return undef causes cache disable
                
                return $key;
            },
        });
    }
    
    sub some_where {
        Mojolicious::Plugin::Cache::Lite->set_expire(sub {
            my $cache_timestamp = shift;
            return 1;
        });
    }

=head1 DESCRIPTION

Mojolicious::Plugin::Cache::Lite provides on memory cache mechanism for
mojolicious.

This plugin caches whole response into key-value object and returns it for next
request instead of invoking on_process code. You can specify the cache key by
giving code reference which gets mojolicious controller for argument.

You can also specify one or more expiration conditions for each cache key from
anywhere in your app by giving code references.

In many case, a single page output involves not only one data model and each of
the models may should have own cache expiration conditions. To expire a cache
exactly right timing, the cache itself must know when to expire. The feature of
this class provides the mechanism.

=head1 METHODS

=head2 register

$plugin->register;

Register plugin hooks in L<Mojolicious> application.

=head2 Mojolicious::Plugin::Cache::Lite->set_expire($code_ref)

This appends a code reference for cache expiration control. 
    
    package Model::NewsRelease;
    
    my $sqlite_file = 'news_release.sqlite';
    
    sub list {
        ...
        
        Mojolicious::Plugin::Cache::Lite->set_expire(sub {
            my $cache_timestamp = shift;
            return $cache_timestamp - (stat($sqlite_file))[9] > 0;
        });
        ...
    }

=head1 AUTHOR

Sugama, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Sugama.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut


lib/Mojolicious/Plugin/CacheLite.pm
package Mojolicious::Plugin::CacheLite;
use strict;
use warnings;
our $VERSION = '0.03';
use base qw/Mojolicious::Plugin/;
use Mojo::Cache::Extended;
use Time::HiRes qw(time);

    our $_EXPIRE_CODE_ARRAY = [];
    
    sub set_expire {
        my ($class, $code) = @_;
        push @$_EXPIRE_CODE_ARRAY, $code;
    }
    
    sub register {
        my ( $self, $app, $conf ) = @_;
        
        my $keygen = $conf->{keygen} || sub {
            shift->req->url->to_abs->to_string;
        };
        
        my $cache = Mojo::Cache::Extended->new(
            size_of => sub {shift->content->body_size}
        );
        
        if ($conf->{max_size}) {
            $cache->max_size($conf->{max_size});
        }
        
        if ($conf->{max_keys}) {
            $cache->max_keys($conf->{max_keys});
        }
        
        my $on_process_org = $app->on_process;
        
        $app->on_process(sub {
            
            my ($app, $c) = @_;
            
            if ($c->req->method eq 'GET' && (my $key = $keygen->($c))) {
                
                if (my $res = $cache->get($key)) {
                    $app->log->debug("serving from cache for $key");
                    $c->tx->res($res);
                    $c->rendered;
                    return;
                }
                
                local $_EXPIRE_CODE_ARRAY;
                
                my $ts_s = time;
                
                $on_process_org->($app, $c);
                
                if (time - $ts_s > ($conf->{threshold} || 0)) {
                    my $code = $c->res->code;
                    if ($code && $code == 200) {
                        $app->log->debug("storing in cache for $key");
                        $cache->set($key, $c->res);
                        for my $code (@$_EXPIRE_CODE_ARRAY) {
                            $cache->set_expire($key, $code);
                        }
                    }
                }
            } else {
                $on_process_org->($app, $c);
            }
        });
    }

1;

__END__

=head1 NAME

Mojolicious::Plugin::CacheLite - On memory cache plugin [ALPHA]

=head1 SYNOPSIS

    use Mojolicious::Plugin::CacheLite;
    
    sub startup {
        my $self = shift;
        
        $self->plugin('cache_lite');
        
        or
        
        $self->plugin(cache_lite => {
            max_size => 1000000, # byte
            max_keys  => 100,    
            threshold => 0.08,   # second
            keygen => sub {
                my $c = shift;
                
                # generate key here maybe with $c
                # return undef causes cache disable
                
                return $key;
            },
        });
    }
    
    sub some_where {
        Mojolicious::Plugin::CacheLite->set_expire(sub {
            my $cache_timestamp = shift;
            return 1;
        });
    }

=head1 DESCRIPTION

Mojolicious::Plugin::CacheLite provides on memory cache mechanism for
mojolicious.

This plugin caches whole response into key-value object and returns it for next
request instead of invoking on_process code. You can specify the cache key by
giving code reference which gets mojolicious controller for argument.

You can also specify one or more expiration conditions for each cache key from
anywhere in your app by giving code references. In many case, a single page
output involves not only one data model and each of the models may should have
own cache expiration conditions. To expire a cache exactly right timing,
the cache itself must know when to expire. The feature of this class provides
the mechanism.

=head1 OPTIONS

=head2 keygen => code reference [optional]

Key generator for cache entries. This must be given in code reference.
The following is the default.

    $self->plugin(cache_lite => {
        keygen => sub {
            $c = shift; ## mojolicious controller
            return $c->req->url->to_abs->to_string;
        }
    });

returning undef causes both cache generation and reference disabled.

=head2 max_size => number [optional]

Maximum byte length for total of body lengths of cache. Default is 5000000.
Since it only measures body length, this value must be considered for rough
limitation for memory size.

    $self->plugin(cache_lite => {max_size => 5000000});

=head2 max_keys => number [optional]

Maximum number of cache keys, defaults to 100.

    $self->plugin(cache_lite => {max_keys => 100});

=head2 threshold => number [optional]

Threshold time interval for page generation to activate cache generation.
You can give it a floating number of second. This plugin measures how long
the page generation spent and compares to the threshold. 

    $self->plugin(cache_lite => {threshold => 0.8});

=head1 METHODS

=head2 register

$plugin->register;

Register plugin hooks in L<Mojolicious> application.

=head2 Mojolicious::Plugin::CacheLite->set_expire($code_ref)

This appends a code reference for cache expiration control. 
    
    package Model::NewsRelease;
    
    my $sqlite_file = 'news_release.sqlite';
    
    sub list {
        ...
        
        Mojolicious::Plugin::CacheLite->set_expire(sub {
            my $cache_timestamp = shift;
            return $cache_timestamp - (stat($sqlite_file))[9] > 0;
        });
        ...
    }

=head1 AUTHOR

Sugama, E<lt>sugama@jamadam.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Sugama.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=cut
